src/simulator/services/persistent_state.py

```diff
@@ -0,0 +1,184 @@
+            self.colours[n] = DynamicColour(colour, n, self.__colour_callback, visible)
+
+    def _to_json(self) -> Dict[str, Any]:
+        data = {}
+        dc = data["colours"] = {}
+        for n, c in self.colours.items():
+            dc[n] = {}
+            dc[n]["colour"] = tuple(c.colour)
+            dc[n]["visible"] = c.visible
+        return data
+
+    def is_effective(self) -> bool:
+        return self.index == PersistentStateView.EFFECTIVE_VIEW
+
+class PersistentState(Service):
+    file_name: str
+
+    MAX_VIEWS: Final[int] = 6
+    effective_view: PersistentStateView
+    views: List[PersistentStateView]
+    __view_idx: int
+
+    _save_task: Optional['Task']
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **exclude_from_dict(kwargs, ["file_name"]))
+        self.file_name = kwargs["file_name"] if "file_name" in kwargs else ".pathbench.json"
+
+        self._save_task = None
+
+        self.reset(False)
+        self.load()
+        self.save()
+
+    def reset(self, save: bool = True) -> None:
+        self.effective_view = PersistentStateView(self._services, self, PersistentStateView.EFFECTIVE_VIEW)
+        self.views = [PersistentStateView(self._services, self, i) for i in range(self.MAX_VIEWS)]
+        self.__view_idx = 0
+        if save:
+            self.save()
+
+    def load(self) -> None:
+        def check(req: bool) -> None:
+            if not req:
+                raise RuntimeError
+
+        if os.path.isfile(self.file_name):
+            self._services.debug.write("Loading state from '{}'".format(self.file_name), DebugLevel.BASIC)
+            with open(self.file_name, 'r') as f:
+                try:
+                    jdata = json.load(f)
+                    jidx = jdata["view_index"]
+                    check(jidx >= 0 and jidx < self.MAX_VIEWS)
+                    jviews = jdata["views"]
+                    for v in range(self.MAX_VIEWS):
+                        self.views[v]._from_json(jviews[v])
+                    self.view_idx = jidx
+                except Exception:
+                    print("{} is corrupted".format(self.file_name))
+                    self.reset()
+        else:
+            self._services.debug.write("'{}' not found, falling back to default state data".format(self.file_name), DebugLevel.BASIC)
+
+    def save(self) -> None:
+        self._services.debug.write("Saving state to '{}'".format(self.file_name), DebugLevel.BASIC)
+        data = {}
+        data["view_index"] = self.view_idx
+        data["views"] = [self.views[v]._to_json() for v in range(self.MAX_VIEWS)]
+        with open(self.file_name, 'w') as f:
+            json.dump(data, f, sort_keys=True, indent=4)
+
+        if self._save_task is not None:
+            self._services.graphics.window.taskMgr.remove(self._save_task)
+            self._save_task = None
+
+    def schedule_save(self, delay: float = 0) -> None:
+        if delay < 0:
+            return
```

Comment:
Hmm, the normal way a negative delay is handled is to treat it as no delay, which would simplify the logic here a little.

Reply:
This way we make saving optional with a single argument, otherwise we have to pollute every function with extra args. Doing the check in schedule_save ensures uniform behaviours across all function -> less surprises

Reply:
It's not "pollution" if it makes your code more readable. But you have a good point, as long as it's consistent then it seems alright.